pinc/ProofProject.inc

```diff
@@ -158,3 +179,169 @@ class TextChecker
+                $this->append_word_to_array($word, self::ACCEPTED_WORD);
+            } elseif (in_array($word, $uncommonScriptWords)) {
+                $this->process_uncommon_script($word, $uncommonScriptMap);
+            }
+        }
+        // append any remaining good text
+        $this->append_good(strlen($this->page_text));
+        $this->append_good_char_buffer();
+        return [
+            "wc_array" => $this->wc_array,
+            "scripts" => $uncommonScripts,
+        ];
+    }
+
+    // append text up to $wordIndex
+    private function append_good($wordIndex)
+    {
+        // look for punctuation in this text
+        while ($this->text_index < $wordIndex) {
+            // get each grapheme in turn
+            $character = grapheme_extract($this->page_text, 1, GRAPHEME_EXTR_COUNT, $this->text_index, $this->text_index);
+            // punctuation except in an HTML entity (aka: &amp; &lt; &gt;)
+            // (text_index has already advanced beyond the character)
+            if (in_array($character, $this->puncArray) && !($character == ';' && preg_match("/&\w+;$|&#\d+;$/", substr($this->page_text, 0, $this->text_index)))) {
+                $this->append_good_char_buffer();
+                $this->wc_array[] = [$character, self::PUNCTUATION];
+            } else {
+                $this->good_char_buffer .= $character;
+            }
+        }
+    }
+
+    private function append_word_to_array($word, $type)
+    {
+        $this->append_good_char_buffer();
+        $this->wc_array[] = [$word, $type];
+        $this->text_index += strlen($word);
+    }
+
+    private function process_uncommon_script($word, $scriptMap)
+    {
+        global $common_unicode_scripts;
+
+        foreach (split_multiscript_string_mb($word) as [$chunk, $chunk_script]) {
+            if (in_array($chunk_script, $common_unicode_scripts)) {
+                $this->good_char_buffer .= $chunk;
+            } else {
+                $this->append_good_char_buffer();
+                $this->wc_array[] = [$chunk, self::UNCOMMON_SCRIPT_BASE + $scriptMap[$chunk_script]];
+            }
+        }
+        $this->text_index += strlen($word);
+    }
+
+    private function append_good_char_buffer()
+    {
+        if ($this->good_char_buffer !== "") {
+            $this->wc_array[] = [$this->good_char_buffer, self::GOOD_TEXT];
+            $this->good_char_buffer = "";
+        }
+    }
+}
+
+function make_grapheme_array($string)
+{
+    $next = 0;
+    $maxbytes = strlen($string);
+    $graphemes = [];
+    while ($next < $maxbytes) {
+        $grapheme = grapheme_extract($string, 1, GRAPHEME_EXTR_COUNT, $next, $next);
+        if (false === $grapheme) {
+            throw new Exception("grapheme_extract failure");
+        }
+        $graphemes[] = $grapheme;
+    }
+    return $graphemes;
+}
+
+/**
+ * Split a string into chunks on Unicode boundaries.
+ */
+function split_multiscript_string_mb($string)
```

Comment:
It would be cleaner if we just modified `split_multiscript_string()` to return the `$chunk_script` like you need and adapted the one caller in `spellcheck_text.inc` instead of creating a whole new function that does almost exactly the same thing.

Reply:
done this.